### PR TITLE
Return diversity info in get paths response

### PR DESCRIPTION
**Pull request by sergii-iakovenko**

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/YFlowMapper.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ protected Collection<Flow> getDiverseWithFlow(Flow flow, FlowRepository flowRepo
         return flow.getDiverseGroupId() == null ? Collections.emptyList() :
                 flowRepository.findByDiverseGroupId(flow.getDiverseGroupId()).stream()
                         .filter(diverseFlow -> !flow.getFlowId().equals(diverseFlow.getFlowId())
-                                || (flow.getYFlowId() != null && !flow.getYFlowId().equals(diverseFlow.getYFlowId())))
+                                && (flow.getYFlowId() == null || !flow.getYFlowId().equals(diverseFlow.getYFlowId())))
```
Comment (rozdy) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/YFlowMapper.java:109]: Maybe split into two filter calls to increase readability?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/yflow/YFlowReadService.java [hunk 2]
```diff
@@ -114,35 +121,59 @@ public YFlowPathsResponse getYFlowPaths(@NonNull String yFlowId) throws FlowNotF
...
-                            FlowProtectedPathDto.builder();
-                    if (flow.getProtectedForwardPath() != null) {
-                        if (!flow.isOneSwitchFlow()) {
-                            protectedForwardPaths.add(flow.getProtectedForwardPath());
-                        }
-                        protectedDtoBuilder.forwardPath(
-                                FlowPathMapper.INSTANCE.mapToPathNodes(flow, flow.getProtectedForwardPath()));
-
+                    protectedDtoBuilder = buildFlowProtectedPathDto(flow);
+                    if (flow.getProtectedForwardPath() != null && !flow.isOneSwitchFlow()) {
```
Comment (rozdy) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/yflow/YFlowReadService.java:136]: Is it possible to create protected single-switch flow?

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java [hunk 3]
```diff
@@ -85,20 +96,73 @@ public abstract class YFlowMapper {
     public abstract YFlowSharedEndpointEncapsulation toYFlowSharedEndpointEncapsulation(FlowEndpoint endpoint);
 
     @Mapping(target = "sharedPath", source = "sharedPath", resultType = YFlowPath.class)
+    @Mapping(target = "subFlowPaths",
+            expression = "java(toSubFlowPaths(source.getSubFlowPaths(), source.getDiverseWithFlows()))")
     public abstract YFlowPaths toYFlowPaths(YFlowPathsResponse source);
 
+    protected List<SubFlowPath> toSubFlowPaths(List<FlowPathDto> subFlowPaths,
+                                               Map<String, List<FlowPathDto>> diverseWithFlows) {
+        log.error("toSubFlowPaths: {} / {}", subFlowPaths, diverseWithFlows);
```
Comment (rozdy) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java:105]: Why error?

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java [hunk 4]
```diff
@@ -85,20 +96,73 @@ public abstract class YFlowMapper {
...
+    @Mapping(target = "forward", source = "source.forwardPath")
+    @Mapping(target = "protectedPath",
+            expression = "java(toYFlowProtectedPath(source.getProtectedPath(), diverseWithFlows))")
+    @Mapping(target = "diverseGroup", expression = "java(toDiverseGroupPayload(source, diverseWithFlows))")
+    public abstract SubFlowPath toSubFlowPath(FlowPathDto source, List<FlowPathDto> diverseWithFlows);
+
+    protected DiverseGroupPayload toDiverseGroupPayload(FlowPathDto source, List<FlowPathDto> diverseWithFlows) {
+        OverlappingSegmentsStats segmentsStats = source.getSegmentsStats();
+        if (segmentsStats != null) {
+            log.error("toDiverseGroupPayload: {} / {}", source, diverseWithFlows);
```
Comment (rozdy) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java:134]: Same here.

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java [hunk 5]
```diff
@@ -85,20 +96,73 @@ public abstract class YFlowMapper {
...
-    @Mapping(target = "flowId", source = "id")
-    @Mapping(target = "reverse", source = "reversePath")
-    @Mapping(target = "forward", source = "forwardPath")
-    public abstract SubFlowPath toSubFlowPath(FlowPathDto source);
+    protected DiverseGroupPayload toDiverseGroupPayload(FlowProtectedPathDto source,
+                                                        List<FlowPathDto> diverseWithFlows) {
+        if (source != null) {
+            OverlappingSegmentsStats segmentsStats = source.getSegmentsStats();
+            if (segmentsStats != null) {
+                log.error("toDiverseGroupPayload: {} / {}", source, diverseWithFlows);
```
Comment (rozdy) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java:149]: And here.

**Review by rozdy — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/YFlowMapper.java (hunk 1 above)
Comment (sergii-iakovenko) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/YFlowMapper.java]: Agree. Reformatted.

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java (hunk 4 above)
Comment (sergii-iakovenko) [reply src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java]: Removed.

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java (hunk 5 above)
Comment (sergii-iakovenko) [reply src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/YFlowMapper.java]: Removed.

**Pull request merged**